### DISCUSSION TITLE
HuskyLens Test with AprilTag Array (ex DFRobot)

### DISCUSSION THREAD

**Discussion by baqwas** [FIRST-Tech-Challenge/FtcRobotController#765] [Q&A] [answered]
Hello!

Finally made it to the HuskyLens OpMode!

Based on the suggestion in the FTC docs, I tried the DFRobot AprilTag 7x5 array. The SDK AprilTagEasy OpMode identifies most of the tags (with appropriate IDs where applicable). I'm sure that better lighting and positioning of the array will complete the full sweep. However, the SDK HuskyLensBasic identifies only six. Presumably, (I'm out on a limb here) these are the only ones in the array that are valid for FTC purposes?

My bigger question is how do I know which is which? All IDs are reported as 0 but with distinguishing pixel values for the outline.

Of course, I need to do a lot more work on placement of the array in front of the camera before jumping to conclusions but some quick guidance from experienced practitioners will save me some cycles. The COLOR_RECOGNITION algorithm is working fine. Thanks for making the starter/Concept code so easy to tweak.

Regards.

**Comment by cmacfarl** [comment, marked as answer]
The HuskyLens can "detect" AprilTags.  That doesn't mean that it can decode the ID that the AprilTag represents.

You can "assign" IDs to AprilTags through it's learning mechanism.  See https://wiki.dfrobot.com/HUSKYLENS_V1.0_SKU_SEN0305_SEN0336#target_20

It is returning 6 elements, all reporting 0 zero, because it recognizes 6 items as AprilTags, but none of them have been "trained", and so from the HuskyLens firmware's perspective, they all have ID 0.  As to why only 6?  That's an internal SDK restriction designed to simplify implementation and usage, but still satisfy all requirements of the game.  

See the documentation below.

```
/**
 * Driver for HuskyLens Vision Sensor.
 *
 * {@link HuskyLens} provides support for the DFRobot HuskyLens Vision Sensor.
 *
 * @see <a href="https://wiki.dfrobot.com/HUSKYLENS_V1.0_SKU_SEN0305_SEN0336">HuskyLens</a>
 *
 * Limitations: Both Block and Arrow array sizes are capped at 6 elements.  The
 *   call to blocks() will return all Blocks irrespective of ID, capped at 6.  The
 *   call to blocks(int id) will return all Blocks with the given id, capped at 6.
 *   The latter may be useful if you have too many returned objects.  Similarly with
 *   arrows().
 */
```

**Reply by Naruto3446** [reply]
Where is this information used? What website do I use this? A little outside help would be appreciated.

**Reply by cmacfarl** [reply]
It is hard to answer your question without more information on what you are trying to do.

> Where is this information used? 

What information?  An ID from a recognized AprilTag?  That's entirely dependent upon your application.  This is the FTC SDK repository, so assuming you are programming a robot, how you use the ID is entirely up to you.

> What website do I use this?

I don't understand the question.  https://www.dfrobot.com/product-1922.html is the website that describes the HuskyLens.  You use the device, in the context of this repository on an FTC robot.

For further help please completely describe exactly what you want to do.

**Reply by texasdiaz** [reply]
There's also a really good article on ftc-docs on using the HuskyLens that is still very applicable to this year's game.
https://ftc-docs.firstinspires.org/en/latest/devices/huskylens/huskylens.html

**Comment by baqwas** [comment]
Thanks for the clarification, @cmacfarl. Skipping comments is my personal weakness. I didn't understand the full depth of these until you clarified.

Regards.
